### PR TITLE
feat: Add ERC-2258 functionality to StakeLocker (SC-2089)

### PR DESCRIPTION
**Pull request by michaelelliot**
# Description

# Integrations Checklist

- [ ] Have any function signatures changed? If yes, outline below.
- [ ] Have any features changed or been added? If yes, outline below.
- [ ] Have any events changed or been added? If yes, outline below.
- [ ] Has all documentation been updated?

# Changelog

### CODE REVIEW

File: contracts/test/PoolCustodial.t.sol [hunk 1]
```diff
@@ -236,6 +236,39 @@ contract PoolCustodialTest is TestUtil {
...
+        depositAmt *= WAD;
+        custodyAmt *= WAD;
+
+        assertEq(pool.custodyAllowance(address(fez), address(custodian)), 0);
+        assertEq(pool.totalCustodyAllowance(address(fez)),                0);
+
+        fez.increaseCustodyAllowance(address(pool), address(custodian), custodyAmt);
+
+        assertEq(pool.custodyAllowance(address(fez), address(custodian)), custodyAmt, "Fez gives custody to custodian");
+        assertEq(pool.totalCustodyAllowance(address(fez)), custodyAmt, "Total custody allowance goes up");
```
Comment (lucas-manuel) [inline contracts/test/PoolCustodial.t.sol:259]: Remove error messages

File: contracts/test/PoolCustodial.t.sol [hunk 2]
```diff
@@ -236,6 +236,39 @@ contract PoolCustodialTest is TestUtil {
...
+
+        uint256 withdrawableAmt = (depositAmt - custodyAmt) / WAD * USD;
```
Comment (lucas-manuel) [inline contracts/test/PoolCustodial.t.sol:261]: `* USD / WAD`

File: contracts/test/PoolCustodial.t.sol [hunk 3]
```diff
@@ -236,6 +236,39 @@ contract PoolCustodialTest is TestUtil {
         assertEq(pool.balanceOf(address(fox)), transferableAmt);
     }
 
+    function test_custody_and_withdraw(uint256 depositAmt, uint256 custodyAmt) public {
+        Custodian custodian = new Custodian();
+
+        depositAmt = constrictToRange(depositAmt, 100, 1E9,            true);  // $1 - $1b
+        custodyAmt = constrictToRange(custodyAmt,  40, depositAmt / 2, true);  // $1 - half of deposit
```
Comment (lucas-manuel) [inline contracts/test/PoolCustodial.t.sol:243]: Doesnt need to be half for this test

File: contracts/test/PoolCustodial.t.sol [hunk 4]
```diff
@@ -236,6 +236,39 @@ contract PoolCustodialTest is TestUtil {
...
+
+        mintFundsAndDepositIntoPool(fez, pool, depositAmt * USD, depositAmt * USD);
+
+        pat.setLockupPeriod(address(pool), 0);
+
+        // Convert all amounts to WAD, USD not needed for the rest of the test
+        depositAmt *= WAD;
+        custodyAmt *= WAD;
```
Comment (lucas-manuel) [inline contracts/test/PoolCustodial.t.sol:251]: We actually shouldnt do this in this test as we need this USD value

File: contracts/test/StakeLockerCustodial.t.sol [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+        setUpTokens();
+        setUpOracles();
+        setUpFactories();
+        setUpCalcs();
+        setUpActors();
+        createBalancerPool();
+        transferBptsToPoolDelegateAndStakers();
+        setUpLiquidityPool();
+    }
+
```
Comment (lucas-manuel) [inline contracts/test/StakeLockerCustodial.t.sol:24]: Missing three tests in this file

**Review by lucas-manuel — CHANGES_REQUESTED** [review]

**Review by deluca-mike — DISMISSED** [review]

File: contracts/Pool.sol [hunk 6]
```diff
@@ -479,8 +480,8 @@ contract Pool is PoolFDT {
     /**
         @dev   `from` and `to` should always be equal in this implementation.
         @dev   This means that the custodian can only decrease their own allowance and unlock funds for the original owner.
-        @param from   Address which holds to Pool FDTs.
-        @param to     Address which going to be the new owner of the `amount` FDTs.
+        @param from   Address which holds the Pool FDTs.
+        @param to     Address which will be the new owner of the `amount` of FDTs.
         @param amount Number of FDTs transferred.
      */
     function transferByCustodian(address from, address to, uint256 amount) external {
```
Comment (satyamakgec) [inline contracts/Pool.sol:487]: So we should emit this 
`emit CustodyAllowanceChanged(from, msg.sender, oldAllowance, newAllowance);` So first parameter should be `tokenHolder` and that is `from/to` in this case while `msg.sender` here is custodian so it should be the second argument.

File: contracts/StakeLocker.sol [hunk 7]
```diff
@@ -253,6 +258,45 @@ contract StakeLocker is StakeLockerFDT, Pausable {
...
+        uint256 oldAllowance = custodyAllowance[from][msg.sender];
+        uint256 newAllowance = oldAllowance.sub(amount);
+
+        require(to == from,             "SL:INVALID_RECEIVER");
+        require(amount != uint256(0),   "SL:INVALID_AMT");
+
+        custodyAllowance[from][msg.sender] = newAllowance;
+        totalCustodyAllowance[from]        = totalCustodyAllowance[from].sub(amount);
+        emit CustodyTransfer(msg.sender, from, to, amount);
+        emit CustodyAllowanceChanged(from, msg.sender, oldAllowance, newAllowance);
```
Comment (satyamakgec) [inline contracts/StakeLocker.sol:297]: Here it is right

**Review by satyamakgec — CHANGES_REQUESTED** [review]

**Review by satyamakgec — APPROVED** [review]

**Pull request merged**